src/normalize.jl

```diff
@@ -0,0 +1,56 @@
+using LinearAlgebra
+
+function LinearAlgebra.normalize(tn::AbstractITensorNetwork; alg="exact", kwargs...)
+  return normalize(Algorithm(alg), tn; kwargs...)
+end
+
+function LinearAlgebra.normalize(alg::Algorithm"exact", tn::AbstractITensorNetwork)
+  norm_tn = norm_sqr_network(tn)
+  log_norm = logscalar(alg, norm_tn)
+  tn = copy(tn)
+  L = length(vertices(tn))
+  c = exp(log_norm / L)
+  for v in vertices(tn)
+    tn[v] = tn[v] / sqrt(c)
+  end
+  return tn
+end
+
+function LinearAlgebra.normalize(
+  alg::Algorithm"bp",
+  tn::AbstractITensorNetwork;
+  (cache!)=nothing,
+  update_cache=isnothing(cache!),
+  cache_update_kwargs=default_cache_update_kwargs(cache!),
+)
```

Comment:
So it seems like a basic design question here is if normalizing should refer to treating the `tn` as a state that should be normalized to 1 or as something such that you want the result of `contract(tn)` to be 1.

It seems reasonable to define it such that `tn` is a state where you want `contract(norm_network(tn))` to be 1 as you do here, however it may be good to write it in terms of an inner function that takes a tensor network and returns a new one where the tensors are scaled such that contracting it is 1. I can't think of a good name for that right now, but for the time being I'll refer to it as `rescale(tn::AbstractITensorNetwork)`, so `scalar(rescale(tn)) == 1` for any input `tn`, and the input has to be a closed network that evaluates to a scalar. Then we can just define `normalize(tn) = ket_network(rescale(norm_network(tn)))` or something like that.

The current implementation feels a bit too "in the weeds" dealing with quadratic forms, bras, kets, etc. and seems like something that could be abstracted and generalized.

Reply:
Also defining a function like `rescale` then would be relevant for other kinds of networks like partition functions, where if you track the normalization factors then that gives the evaluation of the partition function.

Reply:
Relatedly, `rescale(tn::AbstractITensorNetwork)` could be defined in two steps, one where it computes the local scale factors (I think there is already a function for that?) and then a next step where it just divides the factors by those scale factors, so the implementation could be a bit simpler by dividing it into multiple generic steps.

Reply:
Yeah I see what you mean, that's a nice idea to split it apart like that. Will change it to do that

Reply:
Okay I split it apart. I couldn't quite use `rescale` as ubiquitously for the `alg = "bp"` case because the way that gets renormalized is on a partition vertex by partition vertex basis (so that all the free energy terms are 1). But I was still able to simplify to a reasonable degree.